### PR TITLE
Feature 15/seb - Telestation

### LINKED ISSUE #15: Feature: Telestation
Telestation is a feature that will allow a player at their own discretion to teleport themself back into the action from base after a technical issue, whether it be getting Arma'd or DC'd. This would not require Zeus permission, but would inform the zeus it happened.

General principles:
- Entirely player-driven, no zeus intervention needed
- Player can choose who the teleport target is
- Do not teleport player near enemies or within LoS of friendlies
- Log use of feature (diag_log or inform zeus)

### PR DESCRIPTION
**Pull request by Seb105**
Adds a function which allows players ("caller") to teleport near to their comrades should they encounter technical difficulties. They select a player ("target") to teleport to, and the logic attempts to find them a position according to following criteria:

- Teleport into vehicle cargo if target is in a vehicle and there is room, else find a spot nearby the vehicle
    - If vehicle is full, do not teleport at all if vehicle is moving more than 15kph or is a Helicopter to stop players teleporting and immediately getting left behind
- If target is infantry or teleporting next to vehicle, do not teleport within 10m of a friendly, 50m of an enemy, or in direct line of sight of any unit.

### CODE REVIEW

**Comment by Seb105** [comment]
Tested working in MP.

File: addons/functions/scopebox/fn_scopebox.sqf [hunk 1]
```diff
@@ -1,69 +1,70 @@
 /* ----------------------------------------------------------------------------
```
Comment (milivojm) [inline addons/functions/scopebox/fn_scopebox.sqf:1]: Isn't this not the subject of this enhancement?

File: addons/telestation/fn_teleportUnit.sqf [hunk 2]
```diff
@@ -0,0 +1,48 @@
+params ["_target", "_caller"];
+
+private _teleportSuccess = false;
+private _failPos = [-500, -500, 0];
+
+private _blacklistUnits = allUnits select {_target distance2D _x < 150};
```
Comment (milivojm) [inline addons/telestation/fn_teleportUnit.sqf]: Can you please comment these few lines a bit, just for readability.

File: addons/telestation/fn_ui.sqf [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+params ["_telestation"];
+
+private _fnc_UI = {
+    params ["_caller"];
+
+    [
+        "Telestation",
+        [
+            ["EDIT:MULTI", "The telestation is only to be used in cases of technical problems.", ["", {}, 0], true],
+            ["EDIT:MULTI", "Confirm your use is because of technical difficulties.", ["", {}, 0], true],
```
Comment (milivojm) [inline addons/telestation/fn_ui.sqf:26]: There is no label control?

File: addons/telestation/fn_ui.sqf [hunk 4]
```diff
@@ -0,0 +1,109 @@
+/* ----------------------------------------------------------------------------
```
Comment (milivojm) [inline addons/telestation/fn_ui.sqf:1]: This is up to us, do you think we should have this header on all functions? Personally I like it but I want this decision to be made as collective.

File: addons/telestation/fn_customFindSafePos.sqf [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            {
+                if (_this inArea _x) exitWith {true}; false
+            } forEach _posBlacklist}) exitWith {};
+
+
+        if (_checkUnitBlacklist && {
+            {
+                private _pos = (AGLtoASL _this) vectorAdd [0, 0, 1.8]; // eye level to eye level check
+                private _relDir = _x getRelDir _pos;
+                if ((_relDir < 60 or _relDir > 300) && {[_x, "GEOM"] checkVisibility [eyePos _x, _pos] > 0.01}) exitWith {true}; false // check unit is facing position and position is visible
```
Comment (milivojm) [inline addons/telestation/fn_customFindSafePos.sqf:153]: Maybe introduce preprocessor constants for some constants?

File: addons/functions/scopebox/fn_scopebox.sqf (hunk 1 above)
Comment (Seb105) [reply addons/functions/scopebox/fn_scopebox.sqf:1]: Secret fix to some formatting problems in the code, I did test it still works afterwards though.

File: addons/telestation/fn_ui.sqf (hunk 3 above)
Comment (Seb105) [reply addons/telestation/fn_ui.sqf:26]: Nope! Lots of frustrating testing but this was the best way I found to do it.

File: addons/telestation/fn_ui.sqf (hunk 4 above)
Comment (Seb105) [reply addons/telestation/fn_ui.sqf:1]: I think for functions that should be called by users there should be a header, but it is not necessary for internal functions

File: addons/telestation/fn_customFindSafePos.sqf (hunk 5 above)
Comment (Seb105) [reply addons/telestation/fn_customFindSafePos.sqf:153]: Not sure what it would solve here?

File: addons/telestation/fn_bugDaZeus.sqf [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+if (!isNull getAssignedCuratorLogic player) then { // is reciever a Curator?
+    private _str = format ["%1 %2 attempted to use the Telestation", name _caller, ["unsuccessfully", "successfully"] select _success];
+    
+    if (!isNull findDisplay IDD_RSCDISPLAYCURATOR) then { // are they in the Zeus interface?
+        ["Telestation notification", _str, 30] call BIS_fnc_curatorHint;
+    } else {
+        hint _str;
+    };
+};
```
Comment (freghar) [inline addons/telestation/fn_bugDaZeus.sqf]: No newline, in case you want to fix it.

File: addons/telestation/fn_customFindSafePos.sqf (hunk 5 above)
Comment (freghar) [reply addons/telestation/fn_customFindSafePos.sqf:153]: I'd agree with @Seb105 , these are internal values (degrees of rotation) and not something user-configurable. Similarly, they aren't a well-known constant like `PI` would be, hence I'm fine with leaving them hardcoded.

File: addons/telestation/fn_ui.sqf [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                    private _cfg = missionConfigFile >> "CfgUnitInsignia" >> _insigniaClass;
+                    if (!isClass _cfg) then {
+                        _cfg = campaignConfigFile >> "CfgUnitInsignia" >> _insigniaClass;
+                        if (!isClass _cfg) then {
+                            _cfg = configFile >> "CfgUnitInsignia" >> _insigniaClass;
+                        };
+                    };
+                    
+                    private _texture = getText (_cfg >> "texture");
+                    _texture
```
Comment (freghar) [inline addons/telestation/fn_ui.sqf:54]: Maybe end the function with just `getText (_cfg >> "texture")` (no `_texture` variable)? ... Also, the empty line above reminds one of times and code past. Should it still be there?

**Comment by freghar** [comment]
Small optional fixes can be done, but otherwise LGTM.

File: addons/telestation/fn_customFindSafePos.sqf (hunk 5 above)
Comment (milivojm) [reply addons/telestation/fn_customFindSafePos.sqf:153]: Alright, if not meant to be configured.

**Review by milivojm — APPROVED** [review]

**Pull request merged**